File: Add_two_numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *cur1  = l1, *cur2 = l2, *cur3, *res;
        int flag, sum;
        
        if(cur1 && cur2)
        {
            sum = cur1->val + cur2->val;
            res = cur3 = new ListNode(sum % 10);
            flag = sum / 10;
            cur1 = cur1->next;
            cur2 = cur2->next;
        }
        
        while(cur1 && cur2)
        {
            sum = cur1->val + cur2->val + flag;
            cur3->next = new ListNode(sum % 10);
            flag = sum / 10;
            cur3 = cur3->next;
            cur1 = cur1->next;
            cur2 = cur2->next;
        }
        
        while(cur1)
        {
            sum = cur1->val + flag;
            cur3->next = new ListNode(sum % 10);
            flag = sum / 10;
            cur3 = cur3->next;
            cur1 = cur1->next;
        }
        
        while(cur2)
        {
            sum = cur2->val + flag;
            cur3->next = new ListNode(sum % 10);
            flag = sum / 10;
            cur3 = cur3->next;
            cur2 = cur2->next;
        }
        
        if(flag)
            cur3->next = new ListNode(flag);
        
        return res;
    }
};

File: leetcode/linklist/ Add Two Numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    c1, c2 := l1, l2
    num := 0
    l := 0
    dummyNode := &ListNode{0, nil}
    c := dummyNode
    for;c1 != nil && c2 !=nil;{
        num = c1.Val + c2.Val + l 
        l = num / 10 
        num = num % 10
        c1 = c1.Next
        c2 = c2.Next
        c.Next = &ListNode{num, nil}
        c = c.Next
    }

    for;c1 != nil;{
        num = c1.Val + l 
        l = num / 10
        num = num % 10
        c1 = c1.Next
        c.Next = &ListNode{num, nil}
        c = c.Next
    }

    for;c2 != nil;{
        num = c2.Val + l 
        l = num / 10
        num = num % 10
        c2 = c2.Next
        c.Next = &ListNode{num, nil}
        c = c.Next
    }
    
    if l != 0{
        c.Next = &ListNode{l, nil}
    }
    return dummyNode.Next
}